base/abstractarray.jl

```diff
@@ -3462,8 +3462,11 @@ function ith_all(i, as)
 end
 
 function map_n!(f::F, dest::AbstractArray, As) where F
-    idxs1 = LinearIndices(As[1])
-    @boundscheck LinearIndices(dest) == idxs1 && all(x -> LinearIndices(x) == idxs1, As)
+    idxs1 = eachindex(LinearIndices(As[1]))
+    @boundscheck begin
+        idxs1 = intersect(map(eachindex ∘ LinearIndices, As)...)
+        checkbounds(dest, idxs1)
+    end
     for i = idxs1
```

Comment:
```suggestion
    idxs = mapreduce(LinearIndices, intersect, As)
    checkbounds(dest, inds)
    for i = idxs
```

- I don't see why we need `eachindex` is here
- We should only include bounds-check elision if benchmarks reveal that it makes a difference. In this case, we would also need to add `@propagate_inbounds` to `map!` to make it take effect.
- I think that `mapreduce(LinearIndices, intersect, As)` is more elegant than `intersect(map(LinearIndices, As)...)`, unless there is some performance reason to use the latter.
- If always compute the intersection of indices then `inds` is a better name than `inds1`

Reply:
> the `eachindex` I added for the following reason

Good reason.

> I suppose we should decide if in the first place `@inbounds` should allow skipping the check and iterate through the indices of the first argument? 

`@inbounds` should never expose documented semantics that are otherwise not present. This is because you can't count on bounds checks being removed. For exmaple, Julia could be run with `-check-bound=yes` or the boundschecking might not be inlined into the callsite where `@inbounds` is present. In either case, the bounds checks will remain. Consequently I think the answer to that question is no: `@inbounds` should not change the semantics other than to skip bounds checks.